### PR TITLE
- [FP-2473 - IDE Flows appear as running but nodes don't in tree view…

### PR DESCRIPTION
**Pull request by quirinpa**
- [FP-2473 - IDE Flows appear as running but nodes don't in tree view](https://movai.atlassian.net/browse/FP-2473)

### CODE REVIEW

File: src/editors/Flow/view/Flow.jsx [hunk 1]
```diff
@@ -89,7 +89,6 @@ export const Flow = (props, ref) => {
   );
 
   // State Hooks
-  const [loading, setLoading] = useState(true);
```
Comment (quirinpa) [inline src/editors/Flow/view/Flow.jsx:92]: These change and the next should not be needed

File: src/editors/Flow/view/Core/Graph/GraphBase.js [hunk 2]
```diff
@@ -583,22 +585,28 @@ export default class GraphBase {
...
-  nodeStatusUpdated(nodes) {
-    // Let's wait untill allNodes are loaded to recall this function again
-    // What would happen sometimes is it was called before all nodes were loaded
-    // Causing some nodes to not light up when they should.
-    if (!this.allNodesLoaded)
-      return setTimeout(
-        () => this.nodeStatusUpdated(nodes),
-        NODE_LOAD_CHECK_TIMEOUT
-      );
+  nodeStatusUpdated() {
```
Comment (quirinpa) [inline src/editors/Flow/view/Core/Graph/GraphBase.js:588]: Move code inside nodesLoaded to here.

**Review by PedroCristovao-Movai — CHANGES_REQUESTED** [review]
- Found a regression
image

File: src/editors/Flow/view/Components/FlowTopBar/hooks/useNodeStatusUpdate.js [hunk 3]
```diff
@@ -186,29 +182,13 @@ const useNodeStatusUpdate = (props, robotSelected, viewMode) => {
 
         const running = isOnline && isFlowRunning(robotStatusData.active_flow);
         const runningNodes = running ? getNodesRunning(robotStatusData) : [];
-        const nodeStatus = getNodesToUpdate(
```
Comment (quirinpa) [inline src/editors/Flow/view/Components/FlowTopBar/hooks/useNodeStatusUpdate.js:189]: remove geNodesToUpdate declaration

**Review by PedroCristovao-Movai — CHANGES_REQUESTED** [review]
- The active nodes that we have on the tree view is different from what we get in the websocket

File: src/editors/Flow/view/Components/interface/MainInterface.js [hunk 4]
```diff
@@ -25,6 +31,50 @@ const NODE_PROPS = {
...
+
+  for (let i = 0; i < splits.length - 1; i++) {
+    const split = splits[i];
+    currentObj = currentObj[split] = currentObj[split] || {};
+  }
+
+  currentObj[splits[splits.length - 1]] = value;
+}
+
+function _marks(obj) {
```
Comment (quirinpa) [inline src/editors/Flow/view/Components/interface/MainInterface.js:49]: Thank ChatGPT

File: src/editors/Flow/view/Core/Graph/GraphBase.js [hunk 5]
```diff
@@ -607,18 +615,18 @@ export default class GraphBase {
...
+    for (let i = 0; i < nodePath.length; i++) {
+      const nodeParent = this.getNodeParent(nodePath, i, parent);
 
-    const node = this.nodes.get(nodeName);
-    if (node) node.obj.status = [1, true, "true"].includes(status);
+      if (nodeParent) {
+        this.setNodeStatus(nodeParent, status);
+        parent = nodeParent;
+      }
+    }
```
Comment (manuelnogueiramov) [inline src/editors/Flow/view/Core/Graph/GraphBase.js:629]: From what I understood (bare in mind I'm wrong more often than not :laughing:) this is the whole solution for your [ticket](https://movai.atlassian.net/browse/FP-2473).
I don't think adding this whole extra layer of caching which is prone to errors / bugs, is needed, at least at this point in time. @PedroCristovao-Movai what do you think?

Comment (quirinpa) [reply src/editors/Flow/view/Core/Graph/GraphBase.js:628]: There are two tickets included here. I don't that is enough to fix the first one.

Comment (PedroCristovao-Movai) [reply src/editors/Flow/view/Core/Graph/GraphBase.js:628]: I would like to have a solution that uses the architecture of flow (re-using what we already have), but from what I understood that way is too complex (Paulo tried and couldn't find a solution, in a short time). @manuelnogueiramov I don't mind if you try a different solution.

Comment (PedroCristovao-Movai) [reply src/editors/Flow/view/Core/Graph/GraphBase.js:628]: For me, if it is working properly, and we don't have an alternative solution. I will accept the PR

**Review by PedroCristovao-Movai — CHANGES_REQUESTED** [review]
Start and stop single node is still not working

**Comment by quirinpa** [comment]
Start works tho.

**Comment by quirinpa** [comment]
Fixed

**Comment by quirinpa** [comment]
I think this also fixes Bernardo's bug of the freeze. I can't be sure.

**Comment by quirinpa** [comment]
It no longer crashes my computer to run the flow, and as far as I can see, all the regressions are fixed.

**Review by PedroCristovao-Movai — APPROVED** [review]

**Comment by quirinpa** [comment]
🥳🎊🚀🎉

**Pull request merged**